binaryfusefilter.go

```diff
@@ -6,89 +6,26 @@ import (
 	"math/bits"
 )
 
-type BinaryFuse8 struct {
+type Unsigned interface {
+	~uint8 | ~uint16 | ~uint32
+}
+
+type BinaryFuse[T Unsigned] struct {
 	Seed               uint64
 	SegmentLength      uint32
 	SegmentLengthMask  uint32
 	SegmentCount       uint32
 	SegmentCountLength uint32
 
-	Fingerprints []uint8
-}
-
-func calculateSegmentLength(arity uint32, size uint32) uint32 {
-	// These parameters are very sensitive. Replacing 'floor' by 'round' can
-	// substantially affect the construction time.
-	if size == 0 {
-		return 4
-	}
-	if arity == 3 {
-		return uint32(1) << int(math.Floor(math.Log(float64(size))/math.Log(3.33)+2.25))
-	} else if arity == 4 {
-		return uint32(1) << int(math.Floor(math.Log(float64(size))/math.Log(2.91)-0.5))
-	} else {
-		return 65536
-	}
-}
-
-func calculateSizeFactor(arity uint32, size uint32) float64 {
-	if arity == 3 {
-		return math.Max(1.125, 0.875+0.25*math.Log(1000000)/math.Log(float64(size)))
-	} else if arity == 4 {
-		return math.Max(1.075, 0.77+0.305*math.Log(600000)/math.Log(float64(size)))
-	} else {
-		return 2.0
-	}
-}
-
-func (filter *BinaryFuse8) initializeParameters(size uint32) {
-	arity := uint32(3)
-	filter.SegmentLength = calculateSegmentLength(arity, size)
-	if filter.SegmentLength > 262144 {
-		filter.SegmentLength = 262144
-	}
-	filter.SegmentLengthMask = filter.SegmentLength - 1
-	sizeFactor := calculateSizeFactor(arity, size)
-	capacity := uint32(0)
-	if size > 1 {
-		capacity = uint32(math.Round(float64(size) * sizeFactor))
-	}
-	initSegmentCount := (capacity+filter.SegmentLength-1)/filter.SegmentLength - (arity - 1)
-	arrayLength := (initSegmentCount + arity - 1) * filter.SegmentLength
-	filter.SegmentCount = (arrayLength + filter.SegmentLength - 1) / filter.SegmentLength
-	if filter.SegmentCount <= arity-1 {
-		filter.SegmentCount = 1
-	} else {
-		filter.SegmentCount = filter.SegmentCount - (arity - 1)
-	}
-	arrayLength = (filter.SegmentCount + arity - 1) * filter.SegmentLength
-	filter.SegmentCountLength = filter.SegmentCount * filter.SegmentLength
-	filter.Fingerprints = make([]uint8, arrayLength)
-}
-
-func (filter *BinaryFuse8) getHashFromHash(hash uint64) (uint32, uint32, uint32) {
-	hi, _ := bits.Mul64(hash, uint64(filter.SegmentCountLength))
-	h0 := uint32(hi)
-	h1 := h0 + filter.SegmentLength
-	h2 := h1 + filter.SegmentLength
-	h1 ^= uint32(hash>>18) & filter.SegmentLengthMask
-	h2 ^= uint32(hash) & filter.SegmentLengthMask
-	return h0, h1, h2
-}
-
-func mod3(x uint8) uint8 {
-	if x > 2 {
-		x -= 3
-	}
-	return x
+	Fingerprints []T
 }
 
-// PopulateBinaryFuse8 fills the filter with provided keys. For best results,
+// NewBinaryFuse fills the filter with provided keys. For best results,
 // the caller should avoid having too many duplicated keys.
 // The function may return an error if the set is empty.
-func PopulateBinaryFuse8(keys []uint64) (*BinaryFuse8, error) {
+func NewBinaryFuse[T Unsigned](keys []uint64) (*BinaryFuse[T], error) {
```

Comment:
May be hide generic interface and expose only `PopulateBinaryFuse8(), PopulateBinaryFuse16(), PopulateBinaryFuse32()` publicly to keep API consistent with what we already have.

Reply:
That was a trick with configurable types
```
type MyFuse uint32
...
func main() {
  filter := xorfilter.NewBinaryFuse[MyFuse](keys)
  second(filter)
}

func second(filter *xorfilter.BinaryFuse[MyFuse]) {
  ...
}
```
to change **only one line** when trying to find best binary fuse for use case. Deprecating `PopulateBinaryFuse8`.

To be honest I'm tired of `switch value := value.(type)` usage in established projects. Generics allow to avoid "default" cases as they mostly error, simplify to compiler error.